singularity-eos/eos/singularity_eos.cpp

```diff
@@ -343,327 +350,575 @@ int get_sg_eos( // sizing information
+  ScratchV<double> mass_pte(VAWI("PTE::scratch mass"), scratch_size, nmat);
+  ScratchV<double> sie_pte(VAWI("PTE::scratch sie"), scratch_size, nmat);
+  ScratchV<double> vfrac_pte(VAWI("PTE::scratch vfrac"), scratch_size, nmat);
+  ScratchV<double> temp_pte(VAWI("PTE::scratch temp"), scratch_size, nmat);
+  ScratchV<double> press_pte(VAWI("PTE::scratch press"), scratch_size, nmat);
+  ScratchV<double> rho_pte(VAWI("PTE::scratch rho"), scratch_size, nmat);
+  // declare solver scratch, allocate in the case statement
+  int pte_solver_scratch_size{};
+  ScratchV<double> solver_scratch;
 
   Kokkos::View<int, MemoryTraits<at_int>> res("PTE::num fails");
   Kokkos::View<int, MemoryTraits<at_int>> n_solves("PTE::num solves");
-  // loop over cells
-  portableFor(
-      "PTE", 0, ncell, PORTABLE_LAMBDA(const int &iloop) {
-        // cell offset
-        const int i{offsets_v(iloop) - 1};
-        // get "thread-id" like thing with optimization
-        // for small loops
-        const int32_t token{tokens.acquire()};
-        const int32_t tid{small_loop ? iloop : token};
-        // caching mechanism
-        Real cache[MAX_NUM_LAMBDAS];
-        // calculate total mass in the cell
-        Real mass_sum{0.0};
-        for (int m = 0; m < nmat; ++m)
-          mass_sum += frac_mass_v(i, m);
-        // validate specific volume
-        spvol_v(i) = r_is_inp ? spvol_v(i) : vol_v(i) / mass_sum;
-        // calculate specific internal energy if not already known
-        if (!s_is_inp) {
-          Real eng_sum{0.0};
-          Real vol_sum{0.0};
+  const std::string perf_nums =
+      "[" + std::to_string(nmat) + "," + std::to_string(ncell) + "]";
+  switch (input_int) {
+  case -3: {
+    // T-rho input
+    // set frac_vol = 1/nmat
+    // set rho_i = nmat / spvol * frac_mass_i
+    // iterate PTE solver to obtain internal energies
+    // that results in the input T
+    pte_solver_scratch_size = PTESolverFixedTRequiredScratch(nmat);
+    solver_scratch = ScratchV<double>(VAWI("PTE::scratch solver"), scratch_size,
+                                      pte_solver_scratch_size);
+    const std::string rt_name = "PTE::solve (rho,T) input" + perf_nums;
+    portableFor(
+        rt_name.c_str(), 0, ncell, PORTABLE_LAMBDA(const int &iloop) {
+          // cell offset
+          const int i{offsets_v(iloop) - 1};
+          // get "thread-id" like thing with optimization
+          // for small loops
+          const int32_t token{tokens.acquire()};
+          const int32_t tid{small_loop ? iloop : token};
+          double mass_sum{0.0};
+          // normalize mass fractions
+          // first find the mass sum
+          // also set idxs as the decrement of the eos offsets
+          // to take into account 1 based indexing in fortran
+          for (int m = 0; m < nmat; ++m) {
+            mass_sum += frac_mass_v(i, m);
+            pte_idxs(tid, m) = eos_offsets_v(m) - 1;
+            frac_vol_v(i, m) = 0.0;
+          }
```

Comment:
This can be a function I think since it's used many times

Reply:
done